numba/_dispatcher.cpp

```diff
@@ -27,15 +28,96 @@
  */
 
 #if (PY_MAJOR_VERSION >= 3) && (PY_MINOR_VERSION >= 11)
+#ifndef Py_BUILD_CORE
+    #define Py_BUILD_CORE 1
+#endif
+#include "internal/pycore_frame.h"
+#include "internal/pycore_pyerrors.h"
 
+/*
+ * Code originally from:
+ * https://github.com/python/cpython/blob/deaf509e8fc6e0363bd6f26d52ad42f976ec42f2/Python/ceval.c#L6804
+ */
+static int
+call_trace(Py_tracefunc func, PyObject *obj,
+           PyThreadState *tstate, PyFrameObject *frame,
+           int what, PyObject *arg)
+{
+    int result;
+    if (tstate->tracing) {
+        return 0;
+    }
+    if (frame == NULL) {
+        return -1;
+    }
+    int old_what = tstate->tracing_what;
+    tstate->tracing_what = what;
+    PyThreadState_EnterTracing(tstate);
+    result = func(obj, frame, PyTrace_CALL, NULL);
```

Comment:
There's also nothing that updates the frame's line number.

Reply:
I don't think we need to anymore. The test passes and it's asserting that the lineno is in there.

Reply:
From OOB discussion, @sklam confirms that the line numbers are present already. The CPython Python 3.11 implementation has the frame as a `_PyInterpreterFrame` struct which is internal. I think it's possible to move the code here to work more like the code in CPython, but it will come at the cost of Numba copying in more internal CPython implementation details. Considering copying-in as the alternative, as the current method seems to work, I think it is an acceptable trade off.